components/zcash_protocol/src/value.rs

```diff
@@ -0,0 +1,517 @@
+            Ok(ZatBalance(amount))
+        } else if amount < 0 {
+            Err(BalanceError::Underflow)
+        } else {
+            Err(BalanceError::Overflow)
+        }
+    }
+
+    /// Creates an ZatBalance from a u64.
+    ///
+    /// Returns an error if the amount is outside the range `{0..MAX_MONEY}`.
+    pub fn from_u64(amount: u64) -> Result<Self, BalanceError> {
+        if amount <= MAX_MONEY {
+            Ok(ZatBalance(amount as i64))
+        } else {
+            Err(BalanceError::Overflow)
+        }
+    }
+
+    /// Reads an ZatBalance from a signed 64-bit little-endian integer.
+    ///
+    /// Returns an error if the amount is outside the range `{-MAX_BALANCE..MAX_BALANCE}`.
+    pub fn from_i64_le_bytes(bytes: [u8; 8]) -> Result<Self, BalanceError> {
+        let amount = i64::from_le_bytes(bytes);
+        ZatBalance::from_i64(amount)
+    }
+
+    /// Reads a non-negative ZatBalance from a signed 64-bit little-endian integer.
+    ///
+    /// Returns an error if the amount is outside the range `{0..MAX_BALANCE}`.
+    pub fn from_nonnegative_i64_le_bytes(bytes: [u8; 8]) -> Result<Self, BalanceError> {
+        let amount = i64::from_le_bytes(bytes);
+        ZatBalance::from_nonnegative_i64(amount)
+    }
+
+    /// Reads an ZatBalance from an unsigned 64-bit little-endian integer.
+    ///
+    /// Returns an error if the amount is outside the range `{0..MAX_BALANCE}`.
+    pub fn from_u64_le_bytes(bytes: [u8; 8]) -> Result<Self, BalanceError> {
+        let amount = u64::from_le_bytes(bytes);
+        ZatBalance::from_u64(amount)
+    }
+
+    /// Returns the ZatBalance encoded as a signed 64-bit little-endian integer.
+    pub fn to_i64_le_bytes(self) -> [u8; 8] {
+        self.0.to_le_bytes()
+    }
+
+    /// Returns `true` if `self` is positive and `false` if the ZatBalance is zero or
+    /// negative.
+    pub const fn is_positive(self) -> bool {
+        self.0.is_positive()
+    }
+
+    /// Returns `true` if `self` is negative and `false` if the ZatBalance is zero or
+    /// positive.
+    pub const fn is_negative(self) -> bool {
+        self.0.is_negative()
+    }
+
+    pub fn sum<I: IntoIterator<Item = ZatBalance>>(values: I) -> Option<ZatBalance> {
+        let mut result = ZatBalance::zero();
+        for value in values {
+            result = (result + value)?;
+        }
+        Some(result)
+    }
+}
+
+impl TryFrom<i64> for ZatBalance {
+    type Error = BalanceError;
+
+    fn try_from(value: i64) -> Result<Self, BalanceError> {
+        ZatBalance::from_i64(value)
+    }
+}
+
+impl From<ZatBalance> for i64 {
+    fn from(amount: ZatBalance) -> i64 {
+        amount.0
+    }
+}
+
+impl From<&ZatBalance> for i64 {
+    fn from(amount: &ZatBalance) -> i64 {
+        amount.0
+    }
+}
+
+impl TryFrom<ZatBalance> for u64 {
+    type Error = BalanceError;
+
+    fn try_from(value: ZatBalance) -> Result<Self, Self::Error> {
+        value.0.try_into().map_err(|_| BalanceError::Overflow)
```

Comment:
```suggestion
        value.0.try_into().map_err(|_| BalanceError::Underflow)
```
(The `try_into()` can only fail if `value.0` is negative, which is an underflow.)